### PR TITLE
Feature make nginx storage usable for local development

### LINKED ISSUE #141: Make nginx-storage image usable for local development
In #94 we prepared a script that runs a fake `nginx-storage`-like server in a docker container. This is fine for limited tests but we could actually go a step further and make it easy to run `nginx-storage` locally in a container. Currently all other components (Django server and workers, RabbitMQ, PostgreSQL) can be easily set up in development. nginx-storage is the only one (apart from the Django app) with non-trivial configuration but fortunately it does not depend on Kubernetes so running it locally seems entirely possible after a small refactoring.

1) **Move most of the static configuration from the config map to the container.**
    - Extract Lua code into a separate module with clearly defined functions
        - Remove all template variables (like `{{ upload_report_url }}`) from the function code. Replace them with ordinary Lua variables.
        - The module should load another small Lua module that will be provided by the config map and will contain values for the Lua variables that are meant to replace the template variables.
    - Move `error-pages.conf` to the container.
        - This will affect `nginx-proxy` too. The same source file should be used by both (don't duplicate it).
    - Move the content of the location blocks to a partial config file in the container. The config from the config map should now include this partial config from file instead of having it hard-coded.
        - The `server` block and most of the stuff that's directly in it (`listen` and `gzip_...` directives) should stay in the config map.
    - All the bits specific to Kubernetes should stay in the config map. The idea is for the container to contain all the reusable bits so that we can use them to write different overall configs for different environments but keep the same URLs and functionality.
2) **Create default nginx configuration in the container.**
    - Run it on port 8001.
    - It should be very simple, with no features like HTTPS or even gzip.
    - Assume that all the services it needs (like Conductor and Gatekeeper) are running on 127.0.0.1:8000.
    - It should include the config parts moved to the container.
    - If there's duplication between it and the config map, try to create more partial config files in the container.
3) **Test it in development.**
    - Test this with locally installed RabbitMQ, PostgreSQL, with the Django app running on `gunicorn` with all features enabled and with three local celery workers (one per queue).
        - Please ask @rwrzesien for help with that.
4) **Document steps needed to get the full setup for local development in concent-deployment README.**

The final structure we get should look like this:
- `containers/`
    - `nginx-proxy/`
        - `error-pages.conf`
    - `nginx-storage/`
        - `file-transfer-endpoints.conf`
            - Includes `file-transfer-endpoints.lua`
            - `location /download/`
            - `location /download-auth/`
            - `location = /upload/`
            - `location /upload-auth/`
            - `location @upload-internal`
            - `location /result/`
            - `location @upload-result`
        - `file-transfer-endpoints.lua`
            - Includes `file-transfer-config.lua`
            - `validate_upload_request()`
            - `validate_uploaded_file()`
            - `create_upload_location()`
            - `report_upload()`
            - `return_upload_success()`
        - `file-transfer-config.lua`
            - `upload_report_url` variable set to `http://127.0.0.1:8000`
        - `default.conf`
            - Defines the `server` block
                - No HTTPS and no redirect
                - No gzipping
                - Gatekeeper on `http://127.0.0.1:8000`
            - Includes `error-pages.conf` and `file-transfer-endpoints.conf`
- `kubernetes/config-maps/nginx-storage/`
    - `default.conf.j2`
        - `server` block with full HTTPS, gzipping and the rest of the current config
            - Includes `error-pages.conf` and `file-transfer-endpoints.conf`.
            - Includes `ssl.conf`.
    - `file-transfer-config.lua.j2`
        - `upload_report_url` variable set to a value from `var.yml`
    - `ssl.conf`

### PR DESCRIPTION
**Pull request by bartoszbetka**
Resolves #141

### CODE REVIEW

**Comment by cameel** [comment]
Could you put moving files/parts of code in separate commits? As it is now the pull request is hard to review.

**Comment by bartoszbetka** [comment]
Done.

File: README.md [hunk 1]
```diff
@@ -190,3 +190,52 @@ ansible-playbook deploy.yml                                        \
     --inventory  ../../concent-deployment-values/ansible_inventory \
     --user       $user
 ```
+
+### Build nginx-storage locally
+
+Step-by-step instruction to building and running `nginx-storage` locally.
```
Comment (cameel) [inline README.md]: `instruction` -> `instructions`

File: README.md [hunk 2]
```diff
@@ -190,3 +190,52 @@ ansible-playbook deploy.yml                                        \
     --inventory  ../../concent-deployment-values/ansible_inventory \
     --user       $user
 ```
+
+### Build nginx-storage locally
```
Comment (cameel) [inline README.md]: `Build` -> `Building`

File: README.md [hunk 3]
```diff
@@ -190,3 +190,52 @@ ansible-playbook deploy.yml                                        \
...
+
+Step-by-step instruction to building and running `nginx-storage` locally.
+
+
+- Install require lib:
```
Comment (cameel) [inline README.md]: `Install require lib` -> `Install dependencies`

File: README.md [hunk 4]
```diff
@@ -190,3 +190,52 @@ ansible-playbook deploy.yml                                        \
...
+
+```bash
+apt-get update
+apt-get install make python3 python-pip python3-yaml docker.io
+pip install yasha
+```
+
+
+- Add docker group and your user to it:
```
Comment (cameel) [inline README.md]: Please reword:
``` markdown
- Ensure that your user is in the `docker` group and that this group exists.
  This is necessary to be able to run docker commands without `root` privileges.
  Note that the change will not take effect until you log out of the current shell session.
```

File: README.md [hunk 5]
```diff
@@ -190,3 +190,52 @@ ansible-playbook deploy.yml                                        \
...
+
+
+- Run nginx-storage:
+
+```bash
+docker run                         \
+   --rm                            \
+   --hostname nginx-storage-server \
+   --network host                  \
+   --name nginx-storage            \
```
Comment (cameel) [inline README.md]: Please align the options.

File: README.md [hunk 6]
```diff
@@ -190,3 +190,52 @@ ansible-playbook deploy.yml                                        \
...
+   nginx-storage
+```
+
+
+- Example curl to nginx-storage:
```
Comment (cameel) [inline README.md]: Please reword:
``` markdown
- If everything went OK, you should now be able to reach `nginx-storage` on localhost:
```

File: kubernetes/config-maps/nginx-storage/default.conf.j2 [hunk 7]
```diff
@@ -42,219 +42,9 @@ server {
     # NOTE: OpenResty uses /usr/local/openresty/nginx/, not /etc/nginx/ as the default config location.
     # We have to use absolute path to include files from /etc/nginx/.
     include                  /etc/nginx/snippets/error-pages.conf;
+    include                  /etc/nginx/snippets/file-transfer-endpoints.conf;
 
     set $gatekeeper_api_backend http://gatekeeper.default.svc.cluster.local;
```
Comment (cameel) [inline kubernetes/config-maps/nginx-storage/default.conf.j2]: Does this still work if you set it after the include? Until now it was defined before the endpoints.

File: containers/nginx-storage/install-dependencies.sh.j2 [hunk 8]
```diff
@@ -40,4 +40,7 @@ rm -rf /opt/lua-resty-http
 # Lua libs focusing on input data handling, functional programming and OS path management.
 /usr/local/openresty/luajit/bin/luarocks install penlight
 
+mkdir /srv/storage/
+chown -R nobody:nobody /srv/storage/
+
```
Comment (cameel) [inline containers/nginx-storage/install-dependencies.sh.j2]: This this change related to this pull request?

It should at least be in a separate commit.

File: containers/nginx-storage/file-transfer-endpoints.lua [hunk 9]
```diff
@@ -0,0 +1,173 @@
+
+local file_transfer_operation = {}
```
Comment (cameel) [inline containers/nginx-storage/file-transfer-endpoints.lua]: This should be named the same was as the module (`file_transfer_endpoints`).

File: containers/nginx-storage/file-transfer-endpoints.conf [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+        proxy_pass                $gatekeeper_api_backend/gatekeeper/download-auth/$download_url;
+        proxy_pass_request_body   off;
+        proxy_set_header          Content-Length "";
+    }
+
+
+    location = /upload/ {
+        auth_request /upload-auth/;
+        auth_request_set $checksum $upstream_http_concent_file_checksum;
+        auth_request_set $size     $upstream_http_concent_file_size;
```
Comment (cameel) [inline containers/nginx-storage/file-transfer-endpoints.conf]: Please align this.

File: containers/nginx-storage/default.conf [hunk 11]
```diff
@@ -0,0 +1,12 @@
+error_log logs/error.log notice;
+
```
Comment (cameel) [inline containers/nginx-storage/default.conf]: `default.conf` and `file-transfer-config.lua` should be added in a separate commit. Basically, creating the default, hard-coded config for the image is a separate change and should have its own commit(s).

File: containers/nginx-storage/default.conf [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+server {
+    # HTTP
+    listen 81       default_server;
+
+    # NOTE: OpenResty uses /usr/local/openresty/nginx/, not /etc/nginx/ as the default config location.
+    # We have to use absolute path to include files from /etc/nginx/.
+    include                  /etc/nginx/snippets/error-pages.conf;
```
Comment (cameel) [inline containers/nginx-storage/default.conf]: Please align this properly.

File: containers/nginx-storage/default.conf [hunk 13]
```diff
@@ -0,0 +1,12 @@
+error_log logs/error.log notice;
+
+server {
+    # HTTP
+    listen 81       default_server;
```
Comment (cameel) [inline containers/nginx-storage/default.conf]: This config does not include `file-transfer-endpoints.conf`. How does it provide the `upload/` and `download/` endpoints?

File: README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (bartoszbetka) [reply README.md]: Done.

File: README.md [hunk 14]
```diff
@@ -190,3 +190,52 @@ ansible-playbook deploy.yml                                        \
...
+apt-get install make python3 python-pip python3-yaml docker.io
+pip install yasha
+```
+
+
+- Add docker group and your user to it:
+
+```bash
+sudo groupadd docker
+sudo usermod -aG docker <user>
```
[the remark this answers is not in the document]
Comment (bartoszbetka) [reply README.md]: Done.

File: README.md [hunk 15]
```diff
@@ -190,3 +190,52 @@ ansible-playbook deploy.yml                                        \
...
+```
+
+
+- Go to `concent-deployment` repository and run makefile to build `nginx-storage` image:
+
+```bash
+cd <path_to_concent_deployment_repository>/containers/
```
[the remark this answers is not in the document]
Comment (bartoszbetka) [reply README.md]: Done.

File: containers/nginx-storage/default.conf (hunk 13 above)
[the remark this answers is not in the document]
Comment (bartoszbetka) [reply containers/nginx-storage/default.conf]: Done.

File: containers/nginx-storage/file-transfer-endpoints.conf [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+        content_by_lua_block {
+
+            local file_transfer_operation = require('file-transfer-endpoints')
+
+            file_transfer_operation.validate_uploaded_file()
+            file_transfer_operation.create_upload_location()
+            file_transfer_operation.report_upload()
+            file_transfer_operation.return_upload_success()
+
+
```
[the remark this answers is not in the document]
Comment (bartoszbetka) [reply containers/nginx-storage/file-transfer-endpoints.conf]: Done.

File: containers/nginx-storage/file-transfer-config.lua [hunk 17]
```diff
@@ -0,0 +1,6 @@
+
+local file_transfer_config = {}
```
[the remark this answers is not in the document]
Comment (bartoszbetka) [reply containers/nginx-storage/file-transfer-config.lua]: Done.

File: README.md (hunk 1 above)
Comment (bartoszbetka) [reply README.md]: Done.

File: README.md (hunk 2 above)
Comment (bartoszbetka) [reply README.md]: Done.

File: README.md (hunk 3 above)
Comment (bartoszbetka) [reply README.md]: Done.

File: README.md (hunk 4 above)
Comment (bartoszbetka) [reply README.md]: Done.

File: README.md (hunk 5 above)
Comment (bartoszbetka) [reply README.md]: Done.

File: README.md (hunk 6 above)
Comment (bartoszbetka) [reply README.md]: Done.

File: containers/nginx-storage/install-dependencies.sh.j2 (hunk 8 above)
Comment (bartoszbetka) [reply containers/nginx-storage/install-dependencies.sh.j2]: Yes, those settings allows to upload file to docker image. Without this, upload to nginx-storage locally not working.

File: containers/nginx-storage/file-transfer-endpoints.lua (hunk 9 above)
Comment (bartoszbetka) [reply containers/nginx-storage/file-transfer-endpoints.lua]: Done.

File: containers/nginx-storage/file-transfer-endpoints.conf (hunk 10 above)
Comment (bartoszbetka) [reply containers/nginx-storage/file-transfer-endpoints.conf]: Done.

File: containers/nginx-storage/default.conf (hunk 11 above)
Comment (bartoszbetka) [reply containers/nginx-storage/default.conf]: Done.

File: containers/nginx-storage/default.conf (hunk 12 above)
Comment (bartoszbetka) [reply containers/nginx-storage/default.conf]: Why is not correct now ? This is not change in this pull request.

File: containers/nginx-storage/file-transfer-endpoints.conf (hunk 10 above)
Comment (bartoszbetka) [reply containers/nginx-storage/file-transfer-endpoints.conf]: Why is not correct now ? This is not change in this pull request.

File: containers/nginx-storage/default.conf (hunk 13 above)
Comment (bartoszbetka) [reply containers/nginx-storage/default.conf]: It's include `file-transfer-endpoints.conf`. I don't understand what the problem is ?

File: kubernetes/config-maps/nginx-storage/default.conf.j2 (hunk 7 above)
Comment (bartoszbetka) [reply kubernetes/config-maps/nginx-storage/default.conf.j2]: It's works but I change it to defined before the endpoints.

File: containers/nginx-storage/default.conf (hunk 13 above)
Comment (cameel) [reply containers/nginx-storage/default.conf]: I must have been looking at a specific commit. It looks OK now.

File: containers/nginx-storage/default.conf (hunk 12 above)
Comment (cameel) [reply containers/nginx-storage/default.conf]: Well, if you did not touch this line in this commit (only copied it) then it can stay. But please refactor it later.

This should look like this:
```
include /etc/nginx/snippets/error-pages.conf;
```

File: containers/nginx-storage/file-transfer-endpoints.conf (hunk 10 above)
Comment (cameel) [reply containers/nginx-storage/file-transfer-endpoints.conf]: If you did not touch this line in this commit (only copied it) then it can stay. But please refactor it later.

File: containers/Makefile [hunk 18]
```diff
@@ -1,11 +1,20 @@
...
-    build/nginx-proxy/index.html    \
-
-NGINX_STORAGE_SOURCE :=                            \
-    build/nginx-storage/Dockerfile                 \
-    build/nginx-storage/install-dependencies.sh    \
-    build/nginx-storage/table_utils.lua
+NGINX_PROXY_SOURCE :=                    \
+    build/nginx-proxy/Dockerfile         \
+    build/nginx-proxy/index.html         \
+    build/nginx-proxy/error-pages.conf   \
```
Comment (cameel) [inline containers/Makefile]: Please remove the last `\`.

File: containers/nginx-storage/file-transfer-endpoints.lua [hunk 19]
```diff
@@ -0,0 +1,173 @@
...
+    local table_utils   = require('table_utils')
+
+
+    -- Move the file to its final location.
+    -- Accept the name from Concent-Upload-Path header. We assume that this header was validated by
+    -- the gatekeeper app and is safe to use. In particular, the path must be relative.
+    header_path_with_filename = ngx.req.get_headers()["Concent-Upload-Path"]
+    path_parts                = utils.split('/srv/storage/' .. header_path_with_filename, '/')
+    path                      = ""
+    for index ,path_part in pairs(path_parts) do
```
Comment (cameel) [inline containers/nginx-storage/file-transfer-endpoints.lua]: Please fix style here.

File: containers/nginx-storage/default.conf (hunk 12 above)
Comment (bartoszbetka) [reply containers/nginx-storage/default.conf]: Done.

File: containers/Makefile (hunk 18 above)
Comment (bartoszbetka) [reply containers/Makefile]: Done.

File: containers/nginx-storage/file-transfer-endpoints.lua (hunk 19 above)
Comment (bartoszbetka) [reply containers/nginx-storage/file-transfer-endpoints.lua]: Done.

File: containers/nginx-storage/install-dependencies.sh.j2 (hunk 8 above)
Comment (bartoszbetka) [reply containers/nginx-storage/install-dependencies.sh.j2]: Done.
